udp/udp.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2015 The Chihaya Authors. All rights reserved.
+// Use of this source code is governed by the BSD 2-Clause license,
+// which can be found in the LICENSE file.
+
+// Package udp implements a UDP BitTorrent tracker per BEP 15.
+// IPv6 is currently unsupported as there is no widely-implemented standard.
+package udp
+
+import (
+	"net"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/pushrax/bufferpool"
+
+	"github.com/chihaya/chihaya/config"
+	"github.com/chihaya/chihaya/tracker"
+)
+
+// Server represents a UDP torrent tracker.
+type Server struct {
+	config  *config.Config
+	tracker *tracker.Tracker
+
+	done bool
+}
+
+func (s *Server) serve() error {
+	listenAddr, err := net.ResolveUDPAddr("udp", s.config.UDPListenAddr)
+	if err != nil {
+		return err
+	}
+
+	sock, err := net.ListenUDP("udp", listenAddr)
+	defer sock.Close()
+	if err != nil {
+		return err
+	}
+
+	if s.config.UDPReadBufferSize > 0 {
+		sock.SetReadBuffer(s.config.UDPReadBufferSize)
+	}
+
+	pool := bufferpool.New(1000, 2048)
+
+	for !s.done {
+		buffer := pool.TakeSlice()
+		sock.SetReadDeadline(time.Now().Add(time.Second))
```

Comment:
I'm not sure if this is the best way to allow for a graceful shutdown, feel free to comment.